src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1292 @@
+import ntplib
+import pwd
+import socket
+
+from dns import resolver
+from ldap.controls import SimplePagedResultsControl
+from middlewared.schema import accepts, Bool, Dict, Int, List, Str
+from middlewared.service import job, private, ConfigService, ValidationErrors
+from middlewared.service_exception import CallError
+from middlewared.utils import run
+
+
+class DSStatus(enum.Enum):
+    """
+    Following items are used for cache entries indicating the status of the
+    Directory Service.
+    :FAULTED: Directory Service is enabled, but not HEALTHY.
+    :LEAVING: Directory Service is in process of stopping.
+    :JOINING: Directory Service is in process of starting.
+    :HEALTHY: Directory Service is enabled, and last status check has passed.
+    There is no "DISABLED" DSStatus because this is controlled by the "enable" checkbox.
+    This is a design decision to avoid conflict between the checkbox and the cache entry.
+    """
+    FAULTED = 1
+    LEAVING = 2
+    JOINING = 3
+    HEALTHY = 4
+
+
+class neterr(enum.Enum):
+    JOINED = 1
+    NOTJOINED = 2
+    FAULT = 3
+
+
+class SRV(enum.Enum):
+    DOMAINCONTROLLER = '_ldap._tcp.dc._msdcs.'
+    FORESTGLOBALCATALOG = '_ldap._tcp.gc._msdcs.'
+    GLOBALCATALOG = '_gc._tcp.'
+    KERBEROS = '_kerberos._tcp.'
+    KERBEROSDOMAINCONTROLLER = '_kerberos._tcp.dc._msdcs.'
+    KPASSWD = '_kpasswd._tcp.'
+    LDAP = '_ldap._tcp.'
+    PDC = '_ldap._tcp.pdc._msdcs.'
+
+
+class SSL(enum.Enum):
+    NOSSL = 'off'
+    USESSL = 'on'
+    USETLS = 'start_tls'
+
+
+class ActiveDirectory_DNS(object):
+    def __init__(self, **kwargs):
+        super(ActiveDirectory_DNS, self).__init__()
+        self.ad = kwargs.get('conf')
+        self.logger = kwargs.get('logger')
+        return
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, typ, value, traceback):
+        if typ is not None:
+            raise
+
+    def _get_SRV_records(self, host, dns_timeout):
+        """
+        Set resolver timeout to 1/3 of the lifetime. The timeout defines
+        how long to wait before moving on to the next nameserver in resolv.conf
+        """
+        srv_records = []
+
+        if not host:
+            return srv_records
+
+        r = resolver.Resolver()
+        r.lifetime = dns_timeout
+        r.timeout = r.lifetime / 3
+
+        try:
+
+            answers = r.query(host, 'SRV')
+            srv_records = sorted(
+                answers,
+                key=lambda a: (int(a.priority), int(a.weight))
+            )
+
+        except Exception:
+            srv_records = []
+
+        return srv_records
+
+    def port_is_listening(self, host, port, timeout=1):
+        ret = False
+
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        if timeout:
+            s.settimeout(timeout)
+
+        try:
+            s.connect((host, port))
+            ret = True
+
+        except Exception as e:
+            s.close()
+            raise CallError(e)
```

Comment:
No necessary to change, but I think e could have a finally:

```
try:
    ...
except:
    ...
finally:
    s.close()
```

and remove the other closes